File: frontend/dispatcher/lib/phase/geo.ts
// Distance maths for comparing a captured GPS fix against a precinct's geofence.
// Pure and unit-tested: a sign error here would silently turn a mismatch into a pass.

export interface Coords {
  lat: number
  lng: number
}

export interface Geofence extends Coords {
  radiusMetres: number
}

const EARTH_RADIUS_METRES = 6_371_008.8

function toRadians(degrees: number): number {
  return (degrees * Math.PI) / 180
}

/** Great-circle distance in metres between two fixes. */
export function haversineMetres(a: Coords, b: Coords): number {
  const dLat = toRadians(b.lat - a.lat)
  const dLng = toRadians(b.lng - a.lng)
  const latA = toRadians(a.lat)
  const latB = toRadians(b.lat)

  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(latA) * Math.cos(latB) * Math.sin(dLng / 2) ** 2

  return 2 * EARTH_RADIUS_METRES * Math.asin(Math.sqrt(h))
}

/** Signed distance from the geofence boundary: negative inside, positive outside. Null
 *  when either input is missing — null is NOT zero, which means "exactly on the boundary". */
export function geofenceOffsetMetres(fix: Coords | null, fence: Geofence | null): number | null {
  if (fix === null || fence === null) return null
  return haversineMetres(fix, fence) - fence.radiusMetres
}

/** Distance between the driver's phone and the horse's GPS — evidence in its own right. */
export function separationMetres(a: Coords | null, b: Coords | null): number | null {
  if (a === null || b === null) return null
  return haversineMetres(a, b)
}

/** Narrow a nullable lat/lng pair into Coords, or null if either half is missing. */
export function toCoords(lat: number | null, lng: number | null): Coords | null {
  if (lat === null || lng === null) return null
  return { lat, lng }
}

// Threshold below which a separation prints as whole metres rather than kilometres.
const METRES_PER_KM = 1_000

/** Whole metres under a kilometre, one decimal of km above it — "0.3 km" reads as
 *  rounding noise when 300 m is the gate-vs-yard distinction that matters. */
export function formatSeparation(metres: number): string {
  if (metres < METRES_PER_KM) return `${Math.round(metres)} m`
  return `${(metres / METRES_PER_KM).toFixed(1)} km`
}

File: backend/app/core/geo.py
"""Great-circle distance for comparing a captured GPS fix against a precinct geofence.

Deliberately pure: no DB, no config, no other `app/` imports — a leaf module,
unit-testable in isolation.

Python twin of `frontend/dispatcher/lib/phase/geo.ts`; must stay numerically
aligned or the two could disagree on screen for the same fix.
"""

import math
from decimal import Decimal

# Mean Earth radius in metres (IUGG value); do not change independently of
# frontend/dispatcher/lib/phase/geo.ts.
EARTH_RADIUS_METRES = 6_371_008.8

_MIN_LATITUDE = -90.0
_MAX_LATITUDE = 90.0
_MIN_LONGITUDE = -180.0
_MAX_LONGITUDE = 180.0


def _to_validated_float(value: float | Decimal, *, name: str, low: float, high: float) -> float:
    """Coerce a coordinate to float and enforce it is a finite value within [low, high].

    Rejected loudly rather than silently producing a plausible-looking wrong
    distance (e.g. "12m inside the fence" when the truth is far outside).
    """
    coerced = float(value)

    if math.isnan(coerced) or math.isinf(coerced):
        raise ValueError(f"{name} must be a finite number, got {value!r}")

    if not (low <= coerced <= high):
        raise ValueError(f"{name} must be within [{low}, {high}], got {coerced}")

    return coerced


def haversine_metres(
    lat1: float | Decimal,
    lng1: float | Decimal,
    lat2: float | Decimal,
    lng2: float | Decimal,
) -> float:
    """Great-circle distance in metres between two lat/lng points.

    Accepts float or Decimal (precinct coords are Decimal, tracker fixes are
    float); everything coerces to float before trigonometry, since returning
    a Decimal result would be false precision.

    Bounds are inclusive: latitude in [-90, 90], longitude in [-180, 180].
    +/-180 longitude is legitimate (the antimeridian), not an error.
    """
    lat1_f = _to_validated_float(lat1, name="lat1", low=_MIN_LATITUDE, high=_MAX_LATITUDE)
    lng1_f = _to_validated_float(lng1, name="lng1", low=_MIN_LONGITUDE, high=_MAX_LONGITUDE)
    lat2_f = _to_validated_float(lat2, name="lat2", low=_MIN_LATITUDE, high=_MAX_LATITUDE)
    lng2_f = _to_validated_float(lng2, name="lng2", low=_MIN_LONGITUDE, high=_MAX_LONGITUDE)

    d_lat = math.radians(lat2_f - lat1_f)
    d_lng = math.radians(lng2_f - lng1_f)
    lat1_rad = math.radians(lat1_f)
    lat2_rad = math.radians(lat2_f)

    h = (
        math.sin(d_lat / 2) ** 2
        + math.cos(lat1_rad) * math.cos(lat2_rad) * math.sin(d_lng / 2) ** 2
    )

    # For near-antipodal points, rounding can push sqrt(h) just past 1.0,
    # outside math.asin's domain — clamp rather than crash on valid input.
    root_h = min(1.0, math.sqrt(h))

    return 2 * EARTH_RADIUS_METRES * math.asin(root_h)
